<div align="center">
  <h2 style="color:#1a6eb5; border-bottom: 2px solid #1a6eb5; padding-bottom: 6px;">
    Gestión Cuantitativa de Riesgo de Crédito
  </h2>
  <h4 style="color:#555; margin-top:4px;">Construcción Modelo Scoring de Crédito</h4>
  <p style="font-size:14px;">
    <strong>Universidad de Chile</strong>  | 
    Magíster en Analítica Financiera  | 
    <em>Analítica Financiera</em>
  </p>
  <p style="color:#1a6eb5; font-weight:bold; font-size:13px;">Franco Mansilla Ibáñez</p>
  <br>
  <a href="https://www.francomansilla.com">
    <img src="https://img.shields.io/badge/Web-francomansilla.com-1a6eb5?style=flat-square&logo=globe"/>
  </a>
   
  <img src="https://img.shields.io/badge/pandas-2.x-150458?style=flat-square&logo=pandas&logoColor=white"/>
   
  <img src="https://img.shields.io/badge/numpy-1.x-013243?style=flat-square&logo=numpy&logoColor=white"/>
   
  <img src="https://img.shields.io/badge/scipy-1.x-8CAAE6?style=flat-square&logo=scipy&logoColor=white"/>
   
  <img src="https://img.shields.io/badge/tabulate-0.9-4CAF50?style=flat-square"/>
   
  <img src="https://img.shields.io/badge/plotly-5.x-3F4F75?style=flat-square&logo=plotly&logoColor=white"/>
   
  <img src="https://img.shields.io/badge/statsmodels-0.14-1a6eb5?style=flat-square"/>
   
  <img src="https://img.shields.io/badge/scikit--learn-1.x-F7931E?style=flat-square&logo=scikit-learn&logoColor=white"/>
   
  <img src="https://img.shields.io/badge/mrmr__classif-0.2-8A2BE2?style=flat-square"/>
   
  <a href="https://arxiv.org/abs/1908.05376">
    <img src="https://img.shields.io/badge/Paper-mRMR-red?style=flat-square&logo=arxiv&logoColor=white"/>
  </a>
</div>

  <h2 style="color:#1a6eb5; border-bottom: 2px solid #1a6eb5; padding-bottom: 6px;">


# 1. Instancias 

* Control de verisón de librerías.

In [1]:
#%pip install --upgrade --force-reinstall --no-cache-dir pandas==2.2.2 numpy==1.26.4 scipy==1.13.1 tabulate==0.9.0 plotly==5.23.0 statsmodels==0.14.2 mrmr-selection==0.2.8 scikit-learn==1.5.1
!pip install pandas==2.2.2 numpy==1.26.4 scipy==1.13.1 tabulate==0.9.0 plotly==5.23.0 statsmodels==0.14.2 mrmr-selection==0.2.8 scikit-learn==1.5.1

Looking in indexes: https://fmansii%40bci.cl:****@somosmach.jfrog.io/artifactory/api/pypi/data-analytics-pypi/simple
    python-dotenv (>=0.13.*) ; extra == 'standard'
                   ~~~~~~~^

[notice] A new release of pip is available: 25.2 -> 26.0.1
[notice] To update, run: pip install --upgrade pip


* Importar librerías necesarias.

In [2]:
# Librerías Base de Datos
import pandas as pd
import numpy as np

# Estadística
from scipy.stats import levene
from scipy import stats
from tabulate import tabulate

# Visualización
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import plotly.express as px
import plotly.figure_factory as ff

# Preprocesamiento - Modelamiento
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from mrmr import mrmr_classif
from sklearn.metrics import roc_curve, f1_score, roc_auc_score
from sklearn.base import is_classifier
from sklearn.ensemble import RandomForestClassifier

---
# 2. Base de Datos

En esta sección revisaremos:

* 1. Carga de Base de Datos.
* 2. Split Sample. separamos la muestra en entrenamiento (70%) y validación (30%).

---

**1. Carga de Base de Datos**

In [3]:
file_path = '/Users/francomansilla/Library/CloudStorage/GoogleDrive-franco.andres.mansilla@gmail.com/Mi unidad/universidad de chile/Magister Analitica Negocio/Analítica Financiera v2/3. Riesgo Crédito/Código Python v2/'

df_all = pd.read_excel(file_path + 'Base de Datos Python.xlsx', sheet_name='4. BD a Modelar')

print("Dimensión", df_all.shape)
df_all

Dimensión (500, 13)


,id,fecha_corte,default,edad,n_educ,a_emp,a_dir,ing_h,deu_ing,deu_tc,deu_o,flag_casado,a_exp
0,880,2023-02-01,0,43,1.0,23,19,72.0,7.6,1.18,4.29,1.0,NaN
1,979,2024-02-01,0,39,1.0,6,9,61.0,5.7,0.56,2.91,1.0,36.0
2,898,2023-12-01,0,34,NaN,7,15,40.0,6.4,0.95,1.61,1.0,26.0
3,767,2023-08-01,1,40,3.0,5,3,0.0,16.0,NaN,0.00,0.0,41.0
4,915,2023-08-01,1,21,2.0,2,0,20.0,4.5,0.29,0.61,1.0,3.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
495,821,2023-08-01,0,32,2.0,7,10,32.0,4.3,0.43,0.94,0.0,3.0
496,865,2023-03-01,0,50,1.0,8,27,47.0,6.9,0.40,2.84,0.0,29.0
497,659,2023-07-01,0,43,2.0,16,10,83.0,4.1,0.26,3.14,1.0,21.0
498,104,2024-05-01,1,31,1.0,1,6,21.0,2.5,0.28,0.25,NaN,NaN


In [4]:
df = df_all.drop(columns=['id', 'fecha_corte'])
df.columns

Index(['default', 'edad', 'n_educ', 'a_emp', 'a_dir', 'ing_h', 'deu_ing',
       'deu_tc', 'deu_o', 'flag_casado', 'a_exp'],
      dtype='object')

**2. Split Sample**

In [5]:
df_ent, df_val = train_test_split(df, test_size=0.3, random_state=123456)
print(f"Entrenamiento: {len(df_ent)} filas ({len(df_ent)/len(df)*100:.1f}% del total)")
print(f"Validación:    {len(df_val)} filas ({len(df_val)/len(df)*100:.1f}% del total)")

Entrenamiento: 350 filas (70.0% del total)
Validación:    150 filas (30.0% del total)


---- 
# 3. Análisis Calidad de la Información (ACI)

* 1. Identificación y Tratamiento de Valores Atípicos 
* 2. Identificación y Tratamiento de Missing Values
* 3. Índice de Estabilidad de las Variables (CSI) e Information Value (IV)

----

**1.1. ***Identifcación*** variables con datos atípicos**

La idea es usar los estadisticos para poder conocer la base de datos y detectar:
1. **Concentración:** usando los percentiles, por ejemplo, si p5% = p95% entonces la variable tiene una alta concentración + Asimetría.
2. **Datos atípicos:** y frecuencia usando las medidas combinada con mediana + Curtosis. 

In [5]:
def estadistica_descriptiva(df):
    df_num = df.select_dtypes(include='number')
    
    # Estadísticas numéricas
    stats = df_num.describe(percentiles=[0.01, 0.05, 0.5, 0.95, 0.99]).T
    stats = stats[['min', '1%', '5%', 'mean', '50%', 'std', '95%', '99%', 'max']]
    stats.columns = ['Mínimo', 'P1', 'P5', 'Promedio', 'Mediana', 'Desv. Estándar', 'P95', 'P99', 'Máximo']
    stats['Asimetría'] = df_num.skew()
    stats['Curtosis'] = df_num.kurt()
    stats['Variación %'] = ((stats['Promedio'] - stats['Mediana']) / stats['Promedio']).round(4)
    stats['Num. Únicos'] = df_num.nunique()

    return stats.T

resultado = estadistica_descriptiva(df_ent)
display(resultado)

,default,edad,n_educ,a_emp,a_dir,ing_h,deu_ing,deu_tc,deu_o,flag_casado,a_exp
Mínimo,0.000000,20.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
P1,0.000000,21.000000,1.000000,0.000000,0.000000,0.000000,0.749000,0.000000,0.000000,0.000000,0.000000
P5,0.000000,23.000000,1.000000,0.000000,0.000000,16.000000,1.845000,0.070000,0.296000,0.000000,2.000000
Promedio,0.282857,34.520000,1.792398,8.182857,7.865714,38.269006,10.189429,1.144298,2.501802,0.482517,26.275362
Mediana,0.000000,34.000000,1.000000,7.000000,6.000000,32.000000,8.650000,0.780000,1.730000,0.000000,26.000000
Desv. Estándar,0.451032,8.096014,1.042021,6.371115,6.527150,20.713749,6.773175,1.133106,2.205167,0.500570,14.421836
P95,1.000000,49.550000,4.000000,21.000000,21.000000,77.900000,23.710000,3.695500,7.666500,1.000000,48.000000
P99,1.000000,54.000000,5.000000,23.000000,25.510000,100.590000,27.406000,4.723100,9.452700,1.000000,50.000000
Máximo,1.000000,56.000000,6.000000,27.000000,26.000000,113.000000,30.800000,5.440000,9.970000,1.000000,50.000000
Asimetría,0.968404,0.429341,1.440862,0.629430,0.964125,1.143473,0.793890,1.552422,1.476957,0.070342,-0.134339


In [6]:
vars_alta_variacion = resultado.loc['Variación %'][resultado.loc['Variación %'].abs() > 0.15].index.tolist()
print(f"Variables con variación > 15%: {vars_alta_variacion}")

Variables con variación > 15%: ['default', 'n_educ', 'a_dir', 'ing_h', 'deu_ing', 'deu_tc', 'deu_o', 'flag_casado']


- 1.2. **Tratamiento** variables con datos atípicos

La winzorización es una técnica de tratamiento de datos atípicos que consiste en limitar los valores extremos a un rango específico. El método de establecer el piso y el techo se puede usar varios métodos: `percentiles`, `teorema de Chebyshev` entre otros. El umbral es un 15\% como máximo.

In [7]:
vars_alta_variacion = ['a_emp', 'a_dir', 'ing_h', 'deu_ing', 'deu_tc', 'deu_o']

def winsorizar(df_ent, df_val, variables, lower=0.01, upper=0.99):
    df_ent = df_ent.copy()
    df_val = df_val.copy()
    for var in variables:
        p1  = df_ent[var].quantile(lower)
        p99 = df_ent[var].quantile(upper)
        df_ent[var] = df_ent[var].clip(lower=p1, upper=p99)
        df_val[var] = df_val[var].clip(lower=p1, upper=p99)
    return df_ent, df_val

df_ent, df_val = winsorizar(df_ent, df_val, vars_alta_variacion)

print("--- Entrenamiento ---")
display(estadistica_descriptiva(df_ent[vars_alta_variacion]))

print("--- Validación ---")
display(estadistica_descriptiva(df_val[vars_alta_variacion]))

--- Entrenamiento ---


,a_emp,a_dir,ing_h,deu_ing,deu_tc,deu_o
Mínimo,0.000000,0.000000,0.000000,0.749000,0.000000,0.000000
P1,0.000000,0.000000,0.000000,0.773990,0.000000,0.000000
P5,0.000000,0.000000,16.000000,1.845000,0.070000,0.296000
Promedio,8.160000,7.860114,38.182339,10.166343,1.138574,2.498665
Mediana,7.000000,6.000000,32.000000,8.650000,0.780000,1.730000
Desv. Estándar,6.312093,6.511738,20.426871,6.695629,1.113166,2.194957
P95,21.000000,21.000000,77.900000,23.710000,3.695500,7.666500
P99,23.000000,25.260100,100.348100,27.256060,4.701329,9.425739
Máximo,23.000000,25.510000,100.590000,27.406000,4.723100,9.452700
Asimetría,0.584658,0.953941,1.063057,0.742618,1.461525,1.456223


--- Validación ---


,a_emp,a_dir,ing_h,deu_ing,deu_tc,deu_o
Mínimo,0.000000,0.000000,0.000000,0.749000,0.000000,0.000000
P1,0.000000,0.000000,0.000000,0.749000,0.083500,0.000000
P5,0.000000,0.000000,17.000000,1.790000,0.167500,0.243000
Promedio,8.420000,8.950333,41.379388,9.537727,1.239062,2.473438
Mediana,7.000000,8.000000,34.000000,8.450000,0.870000,1.880000
Desv. Estándar,6.319534,6.785560,24.155142,5.902045,1.086341,2.120997
P95,20.550000,21.550000,95.300000,20.185000,3.607500,7.523000
P99,23.000000,25.510000,100.590000,26.031060,4.553000,9.074258
Máximo,23.000000,25.510000,100.590000,27.406000,4.723100,9.452700
Asimetría,0.587663,0.644993,0.944763,0.758930,1.409202,1.454985


- 2.1. **Identifcación** variables con missing values

**Nota.** En este caso, calcules el porcentaje de missing value con respecto a toda la base de datos (**no sobre la df_ent**), dado que por coincidencia del `split sample` pudieron quedar concentrados en el df_ent.


In [8]:
def missing_values(df):
    missing = pd.DataFrame({
        'Cant. Missing': df.isnull().sum(),
        '% Missing': (df.isnull().sum() / len(df) * 100).round(2)
    })
    missing = missing[missing['Cant. Missing'] > 0].sort_values('% Missing', ascending=False)
    return missing

missing_values(df) # Recuerda usar df original para calcular el porcentaje de missing sobre toda la base

,Cant. Missing,% Missing
a_exp,113,22.6
flag_casado,102,20.4
n_educ,13,2.6
deu_tc,12,2.4
ing_h,11,2.2
deu_o,9,1.8


- 2.2. **Tratamiento** variables con missing values

La idea es imputar la variables con missing value respetando la naturaleza de ella. 

* **Si la variables tiene +15\%** de missing value eliminala de la base de datos, pero antes dicotomomízala para conservar la información de que el valor es missing (1: con información, 0: sin información).

* **Si la variables tiene <15\%** de missing value, imputala con la promedio (si es numérica) o con la moda+1 (si es categórica).

In [9]:
def imputar_missing(df_ent, df_val, umbral=0.15):

    df_ent = df_ent.copy()
    df_val = df_val.copy()

    missing_pct = df.isnull().sum() / len(df) # Recuerda usar df original para calcular el porcentaje de missing sobre toda la base

    vars_missing = missing_pct[missing_pct > 0].index.tolist()
    vars_drop    = [v for v in vars_missing if missing_pct[v] >  umbral]
    vars_imputar = [v for v in vars_missing if missing_pct[v] <= umbral]

    # --- Variables con missing > 15%: dummy + drop ---
    for var in vars_drop:
        df_ent[f'{var}_missing'] = df_ent[var].isnull().astype(int)
        df_val[f'{var}_missing'] = df_val[var].isnull().astype(int)
        
    df_ent.drop(columns=vars_drop, inplace=True)
    df_val.drop(columns=vars_drop, inplace=True)
    print(f"Variables eliminadas (missing > {umbral*100:.0f}%): {vars_drop}")

    # --- Variables con missing <= 15%: imputación ---
    for var in vars_imputar:
        if df_ent[var].dtype == 'float64':
            valor = df_ent[var].mean()
            df_ent[var] = df_ent[var].fillna(valor)
            df_val[var] = df_val[var].fillna(valor)
            print(f"{var} (float) imputado con promedio: {valor:.4f}")

        elif df_ent[var].dtype == 'int64':
            valor = df_ent[var].max() + 1
            df_ent[var] = df_ent[var].fillna(valor)
            df_val[var] = df_val[var].fillna(valor)
            print(f"{var} (int) imputado con categoría faltante: {valor}")
    return df_ent, df_val

df_ent, df_val = imputar_missing(df_ent, df_val)

Variables eliminadas (missing > 15%): ['flag_casado', 'a_exp']
n_educ (float) imputado con promedio: 1.7924
ing_h (float) imputado con promedio: 38.1823
deu_tc (float) imputado con promedio: 1.1386
deu_o (float) imputado con promedio: 2.4987


**3. Índice de Estabilidad de las Variables (CSI) e Information Value (IV)**

El **CSI** es una medida que se utiliza para evaluar la estabilidad de una variable a lo largo del tiempo o entre diferentes muestras. Se calcula comparando la distribución de la variable en dos conjuntos de datos, como el conjunto de entrenamiento y el conjunto de validación. 

La formula para calcular el CSI es la siguiente: 
$$CSI = \sum_{i=1}^{n} \frac{(P_{i,train} - P_{i,val})^2}{P_{i,train}}$$

Los rangos de estabilidad son:
- <0.10 → Estable
- 0.10 - 0.25 → Alerta
- '>0.25 → Inestable

El **IV**, por otro lado, es una medida que se utiliza para evaluar la capacidad predictiva de una variable en relación con una variable objetivo (target). Se calcula utilizando la distribución de la variable en relación con la variable objetivo.

La formula para calcular el IV es la siguiente:
$$IV = \sum_{i=1}^{n} (P_{i,good} - P_{i,bad}) \cdot \ln\left(\frac{P_{i,good}}{P_{i,bad}}\right)$$

Los rangos de poder predictivo del IV son:
- <0.02 → Sin poder
- 0.02 - 0.10 → Débil
- 0.10 - 0.30 → Moderado
- '>.30 → Fuerte

In [10]:
def calcular_csi_iv(df_ent, df_val, variables, target, bins=10):

    resultados = []
    for var in variables:
        if var == target:
            continue
        # Crear bins con df_ent
        _, bin_edges = pd.cut(df_ent[var], bins=bins, retbins=True, duplicates='drop')
        bin_edges[0]  = -np.inf
        bin_edges[-1] =  np.inf
        
        # Distribución entrenamiento
        dist_ent = pd.cut(df_ent[var], bins=bin_edges).value_counts(normalize=True).sort_index()
        dist_val = pd.cut(df_val[var], bins=bin_edges).value_counts(normalize=True).sort_index()
        dist = pd.DataFrame({'ent': dist_ent, 'val': dist_val}).fillna(0)
        dist['ent'] = dist['ent'].replace(0, 0.0001)
        dist['val'] = dist['val'].replace(0, 0.0001)
        dist['csi'] = (dist['val'] - dist['ent']) * np.log(dist['val'] / dist['ent'])
        csi_total = dist['csi'].sum()

        # Information Value
        df_bin = df_ent.copy()
        df_bin['bin'] = pd.cut(df_bin[var], bins=bin_edges)
        grouped = df_bin.groupby('bin')[target].agg(['sum', 'count'])
        grouped.columns = ['eventos', 'total']
        grouped['no_eventos'] = grouped['total'] - grouped['eventos']
        total_ev  = grouped['eventos'].sum()
        total_nev = grouped['no_eventos'].sum()
        grouped['dist_ev']  = grouped['eventos'].replace(0, 0.0001)   / total_ev
        grouped['dist_nev'] = grouped['no_eventos'].replace(0, 0.0001) / total_nev
        grouped['woe'] = np.log(grouped['dist_ev'] / grouped['dist_nev'])
        grouped['iv']  = (grouped['dist_ev'] - grouped['dist_nev']) * grouped['woe']
        iv_total = grouped['iv'].sum()
        resultados.append({'Variable': var, 'CSI': round(csi_total, 4), 'IV': round(iv_total, 4)})
    resultado = pd.DataFrame(resultados).sort_values('IV', ascending=False)
    resultado['Estabilidad'] = resultado['CSI'].apply(
        lambda x: 'Estable' if x < 0.1 else ('Alerta' if x < 0.25 else 'Inestable')
    )
    resultado['Poder Predictivo'] = resultado['IV'].apply(
        lambda x: 'Sin poder' if x < 0.02 else ('Débil' if x < 0.1 else ('Moderado' if x < 0.3 else 'Fuerte'))
    )
    return resultado

csi_iv_resultado = calcular_csi_iv(df_ent, df_val, df_ent.select_dtypes(include='number').columns.tolist(), target='default')
display(csi_iv_resultado)


,Variable,CSI,IV,Estabilidad,Poder Predictivo
5,deu_ing,0.1308,0.8533,Alerta,Fuerte
4,ing_h,0.0705,0.4953,Estable,Fuerte
2,a_emp,0.0475,0.4809,Estable,Fuerte
3,a_dir,0.0932,0.2744,Estable,Moderado
1,n_educ,0.0769,0.2603,Estable,Moderado
0,edad,0.0885,0.2266,Estable,Moderado
7,deu_o,0.0162,0.1062,Estable,Moderado
6,deu_tc,0.0753,0.0904,Estable,Débil
9,a_exp_missing,0.0131,0.0013,Estable,Sin poder
8,flag_casado_missing,0.0293,0.0011,Estable,Sin poder


----
# 4. Análisis pre-eliminar

* Siempre es importante conocer las variables en terminos de poder predictivo antes de hacer cualquier modelelo estadístico o de machine learning, para esto se puede usar técnicas univariadas como: ***Information Value*** (IV), **análisis visual** y **pruebas estadísticas**.

* **Limitación**: es un análisis univariado, no considera la interacción entre variables. Pero funciona como mecanismo de eliminación de variables.

En esta sección revisaremos:

* 1. Discriminación visual de las variables.
* 2. Discriminación estadística de las variables.

----

**1. Discriminación Visual**

* Una forma de analizar la capacidad predictiva de las variables es a través de gráficos, permiten visualizar la distribución de las variables en relación con la variable objetivo (target).

In [11]:
grupo_default    = df[df['default'] == 1]
grupo_no_default = df[df['default'] == 0]

fig = make_subplots(rows=1, cols=2, subplot_titles=['deu_tc (no discrimina)', 'deu_ing (discrimina)'])

# --- deu_tc ---
fig.add_trace(go.Histogram(x=grupo_default['deu_tc'],    nbinsx=20, opacity=0.5, name='Default',    histnorm='percent', marker_color='red',  legendgroup='default'),    row=1, col=1)
fig.add_trace(go.Histogram(x=grupo_no_default['deu_tc'], nbinsx=20, opacity=0.5, name='No Default', histnorm='percent', marker_color='blue', legendgroup='no_default'), row=1, col=1)

# --- deu_ing ---
fig.add_trace(go.Histogram(x=grupo_default['deu_ing'],    nbinsx=20, opacity=0.5, name='Default',    histnorm='percent', marker_color='red',  legendgroup='default',    showlegend=False), row=1, col=2)
fig.add_trace(go.Histogram(x=grupo_no_default['deu_ing'], nbinsx=20, opacity=0.5, name='No Default', histnorm='percent', marker_color='blue', legendgroup='no_default', showlegend=False), row=1, col=2)
fig.update_layout(
    barmode='overlay',
    title='Diferencias entre variables que discriminan la variables objetivo (y = default)',
    paper_bgcolor='rgba(255,255,255,1)',
    plot_bgcolor='rgba(255,255,255,1)',
    width=1200,
    height=600
)
fig.update_yaxes(title_text='Frecuencia Relativa (%)', col=1)
fig.update_yaxes(title_text='Frecuencia Relativa (%)', col=2)
fig.show()

**2. Discriminación Estadística**

* Se evaluar la discriminación de las variables sobre la variable objetivo (`default`), se puede usar las pruebas estadística para determinar mediante un nivel de confianza la variable es capaz de discriminar entre los buenos y malos pagadores.

* Para comparar las medias/medianas de la variable según el grupo (`default`), primero es necesario determinar si las varianzas de ambos grupos son iguales o diferentes. Para ello, se utiliza la prueba de varianza de **Fisher**.

In [12]:
def sign_estadistica(df, var_grupo, alpha_varianza=0.05):
    rs_analisis = []
    cat_0 = df[var_grupo].unique()[0]
    cat_1 = df[var_grupo].unique()[1]
    
    cols_numericas = df.select_dtypes(include=[np.number]).columns.tolist()
    for col in cols_numericas:
        if col != var_grupo:
            grupo_0 = df[df[var_grupo] == cat_0][col].dropna()
            grupo_1 = df[df[var_grupo] == cat_1][col].dropna()
            try:
                # Fisher: razón de varianzas F = var1/var2
                var_0 = np.var(grupo_0, ddof=1)
                var_1 = np.var(grupo_1, ddof=1)
                f_stat = var_0 / var_1
                df_0   = len(grupo_0) - 1
                df_1   = len(grupo_1) - 1
                p_value_varianza = 2 * min(
                    stats.f.cdf(f_stat, df_0, df_1),
                    1 - stats.f.cdf(f_stat, df_0, df_1)
                )
                if p_value_varianza >= alpha_varianza:
                    t_stat, p_value = stats.ttest_ind(grupo_0, grupo_1, equal_var=True)
                else:
                    t_stat, p_value = stats.ttest_ind(grupo_0, grupo_1, equal_var=False)
                rs_analisis.append([col, p_value_varianza, np.abs(t_stat), p_value])
            except Exception:
                continue
    t_stats = [(fila[0], fila[2]) for fila in rs_analisis]
    resultado_ordenado = sorted(rs_analisis, key=lambda x: x[2], reverse=True)
    headers = ['Variable', 'P-value Fisher', 'Estadística t', 'P-value']
    return tabulate(resultado_ordenado, headers=headers, floatfmt=".4f"), t_stats

**Fisher (igualdad de varianzas)**
$$H_0: S_1^2 = S_2^2$$
$$H_1: S_1^2 \neq S_2^2$$
$$F = \frac{S_1^2}{S_2^2} \sim F_{(n_1-1,\ n_2-1)}$$
* Si: $p < \alpha \Rightarrow \text{rechazar } H_0 \Rightarrow \text{varianzas distintas} \Rightarrow \text{Welch}$
* Si: $p \geq \alpha \Rightarrow \text{no rechazar } H_0 \Rightarrow \text{varianzas iguales} \Rightarrow \text{T-Student clásico}$

**T-Student (igualdad de medias)**
$$H_0: \mu_1 = \mu_2$$
$$H_1: \mu_1 \neq \mu_2$$
$$t = \frac{\bar{X}_1 - \bar{X}_2}{\sqrt{S_p^2\left(\frac{1}{n_1} + \frac{1}{n_2}\right)}} \sim t_{(n_1+n_2-2)}$$
* Si: $p < \alpha \Rightarrow \text{rechazar } H_0 \Rightarrow \mu_1 \neq \mu_2 \Rightarrow \text{variable discrimina}$
* Si: $p \geq \alpha \Rightarrow \text{no rechazar } H_0 \Rightarrow \mu_1 = \mu_2 \Rightarrow \text{variable no discrimina}$

In [13]:
sign_stats, _ = sign_estadistica(df, 'default', alpha_varianza=0.05)
print(sign_stats)

Variable       P-value Fisher    Estadística t    P-value
-----------  ----------------  ---------------  ---------
deu_ing                0.0096           7.3148     0.0000
a_emp                  0.0652           6.0728     0.0000
a_dir                  0.1195           2.7665     0.0059
ing_h                  0.2754           2.5573     0.0109
deu_o                  0.2283           2.3091     0.0214
deu_tc                 0.3553           2.2874     0.0226
edad                   0.1051           1.9856     0.0476
n_educ                 0.2630           1.8572     0.0639
flag_casado            0.9517           0.7500     0.4537
a_exp                  0.4264           0.3741     0.7085


* La variable que más discrmina los buenos y malos pagadores (`default = 1`) es la variable ratio `deuda_ingreso` con un `estadístico t = 7.3148`

**TAREA**

Cómo aplicarian el análisis de discriminación estadístico para hacer un filtro de variables mediante **coeficientes de correlación**, usando: 

1. Coef. correlación de Pearson (variables numéricas).
2. Coef. correlación de Cramer (variables categóricas).
3. Coef. correlación de Spearman (variables numéricas con alta concentración o asimetría).
4. Coef. correlación de Theil (variables categóricas con alta concentración o asimetría).
5. Coef. correlación de Tau de Kendall (variables numéricas con alta concentración o asimetría).

---
# 5. Modelamiento

Lograr entrenar un modelo estadístico o de machine learning con tal que logre extraer información de las variables dependientes sobre la variable independiente (target) para identificar patrones logrando estimar correctamente los buenos (y = 0) y malos pagadores (y = 1).

 $$ y = f(X) + \epsilon_i $$
 $$ y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + ... + \beta_n X_n + \epsilon_i $$
 $$ default_i = \beta_0 + \beta_1 \cdot deuda\_ingreso + \beta_2 \cdot edad + ... + \beta_n \cdot X_n + \epsilon $$



## 5.1. Funciones de partidas

Se cargan:

* 1. Separación de variables para entrenamiento y validación
* 2. Función de Kolmogorov-Smirnov (KS).
* 3. Función iterativa para mrMR. 

1. Variables para **Entrenamiento** y **Validación**

In [15]:
y_ent = df_ent['default']
x_ent = df_ent.drop(columns=['default'])    

y_val = df_val['default']
x_val = df_val.drop(columns=['default'])

print("Dimensión Entrenamiento:", x_ent.shape)
print("Dimensión Validación:", x_val.shape)
print("Variables de Entrenamiento:", x_ent.columns.tolist())
print("Variables de Validación:", x_val.columns.tolist())


Dimensión Entrenamiento: (350, 10)
Dimensión Validación: (150, 10)
Variables de Entrenamiento: ['edad', 'n_educ', 'a_emp', 'a_dir', 'ing_h', 'deu_ing', 'deu_tc', 'deu_o', 'flag_casado_missing', 'a_exp_missing']
Variables de Validación: ['edad', 'n_educ', 'a_emp', 'a_dir', 'ing_h', 'deu_ing', 'deu_tc', 'deu_o', 'flag_casado_missing', 'a_exp_missing']


2. Métrica a usar **Kolmogorov-Smirnov (KS)**.

In [16]:
# Cálculo de KS
def ks_statistic(y_true, y_pred_proba):
    fpr, tpr, _ = roc_curve(y_true, y_pred_proba)
    ks = np.max(tpr - fpr)
    return ks

3. **Método de MrMR** (Mínima Redundancia - Máxima Relevancia) - [Documentación de MrMR](https://feature-engine.trainindata.com/en/1.8.x/user_guide/selection/MRMR.html)

In [17]:
def ks_vars(estimator, x_ent, y_ent, x_val, y_val, max_vars=2, graph=False, subtitulo=''):
    ks_ent = []
    ks_val = []
    variable_names = []

    for num_vars in range(1, max_vars + 1):
        selected_features = mrmr_classif(x_ent, y_ent, K=num_vars)
        variable_names.append(list(selected_features))

        X_ent_selected = x_ent[selected_features]
        X_val_selected = x_val[selected_features]

        # Para modelos de statsmodels, agrega una constante
        if not is_classifier(estimator):
            X_ent_selected = sm.add_constant(X_ent_selected)
            X_val_selected = sm.add_constant(X_val_selected)
            model = estimator(y_ent, X_ent_selected)
            fitted_model = model.fit()
            y_pred_proba_ent = fitted_model.predict(X_ent_selected)
            y_pred_proba_val = fitted_model.predict(X_val_selected)

        else:
            # Para clasificadores de sklearn
            estimator.fit(X_ent_selected, y_ent)
            y_pred_proba_ent = estimator.predict_proba(X_ent_selected)[:, 1]
            y_pred_proba_val = estimator.predict_proba(X_val_selected)[:, 1]

        ks_ent_stat = ks_statistic(y_ent, y_pred_proba_ent)
        ks_val_stat = ks_statistic(y_val, y_pred_proba_val)

        ks_ent.append(ks_ent_stat)
        ks_val.append(ks_val_stat)

    if graph:
        import plotly.graph_objects as go
        fig = go.Figure()
        fig.add_trace(go.Scatter(x=np.arange(1, max_vars + 1), y=ks_ent, mode='lines+markers', name='Entrenamiento', line=dict(color='black', dash='solid')))
        fig.add_trace(go.Scatter(x=np.arange(1, max_vars + 1), y=ks_val, mode='lines+markers', name='Validación', line=dict(color='black', dash='dash')))
        fig.update_layout(title='Selección de variables usando KS',
                          xaxis_title='Número de variables',
                          yaxis_title='Estadístico KS',
                          legend_title='Conjunto',
                          width=1000, height=500,
                          template='simple_white',
                          title_text='Selección de Features usando Kolmogorov-Smirnov (KS)<br>'+subtitulo)
        fig.show()

    return ks_ent, ks_val, variable_names

## 5.2. Regresión Logística

La regresión logística es un modelo de clasificación binaria que estima la probabilidad de que un evento ocurra dado un conjunto de variables independientes. A diferencia de la regresión lineal, la variable dependiente es dicotómica (0/1), por lo que se utiliza la función sigmoide para acotar la salida entre 0 y 1.

**Modelo**
$$P(Y=1 \mid X) = \frac{1}{1 + e^{-(\beta_0 + \beta_1 X_1 + \beta_2 X_2 + \cdots + \beta_k X_k)}}$$

**Odds y Log-Odds** : ¿Cuántas veces más probable es que ocurra el evento en un grupo (P(Y=1)) versus otro (1 - P(Y=1))?
$$\text{Odds} = \frac{P(Y=1)}{1 - P(Y=1)}$$
$$\text{Logit} = \ln\left(\frac{P(Y=1)}{1-P(Y=1)}\right) = \beta_0 + \beta_1 X_1 + \cdots + \beta_k X_k$$

**Estimación de parámetros (Máxima Verosimilitud)** : Busca los valores de $\beta$ que maximizan la probabilidad de observar los datos que efectivamente ocurrieron.

$$\mathcal{L}(\beta) = \prod_{i=1}^{n} P(Y_i=1)^{y_i} \cdot (1 - P(Y_i=1))^{1-y_i}$$
$$\ell(\beta) = \sum_{i=1}^{n} \left[ y_i \ln P(Y_i=1) + (1-y_i) \ln(1-P(Y_i=1)) \right]$$

En esta sección revisaremos:

* 1. Ejecución del método de MrMR para selección de variables.
* 2. Entrenamiento del modelo de regresión logística.
* 3. Evaluación Metodologica, parte 1: métrica KS, F1, AUC, entre otras.
* 4. Evaluación Metodologica, parte 2: diagnostico de distribución de errores predictivos por variable.
* 5. Evaluación de Negocio.


**1. Ejecución del método de MrMR para selección de variables.**

In [18]:
max_vars = len(x_ent.columns.tolist())
logit_model = sm.Logit
ks_ent_vars_cl, ks_val_vars_cl, var_names = ks_vars(logit_model, x_ent, y_ent, x_val, y_val, max_vars=max_vars, graph=True
                                                                    ,subtitulo='Regresión Logística')

100%|██████████| 1/1 [00:00<00:00, 176.46it/s]


Optimization terminated successfully.
         Current function value: 0.531850
         Iterations 6


100%|██████████| 2/2 [00:00<00:00, 10.65it/s]

Optimization terminated successfully.
         Current function value: 0.490712
         Iterations 6



100%|██████████| 3/3 [00:00<00:00,  6.30it/s]

Optimization terminated successfully.
         Current function value: 0.486632
         Iterations 6



100%|██████████| 4/4 [00:00<00:00,  7.13it/s]

Optimization terminated successfully.
         Current function value: 0.484927
         Iterations 6



100%|██████████| 5/5 [00:00<00:00,  7.79it/s]

Optimization terminated successfully.
         Current function value: 0.480796
         Iterations 6



100%|██████████| 6/6 [00:01<00:00,  4.68it/s]

Optimization terminated successfully.
         Current function value: 0.479819
         Iterations 6



100%|██████████| 7/7 [00:01<00:00,  5.82it/s]

Optimization terminated successfully.
         Current function value: 0.476892
         Iterations 6



100%|██████████| 8/8 [00:01<00:00,  4.72it/s]

Optimization terminated successfully.
         Current function value: 0.476312
         Iterations 6



100%|██████████| 9/9 [00:02<00:00,  4.00it/s]

Optimization terminated successfully.
         Current function value: 0.476270
         Iterations 6



100%|██████████| 10/10 [00:02<00:00,  4.43it/s]


Optimization terminated successfully.
         Current function value: 0.475785
         Iterations 6


**2. Entrenamiento del modelo de regresión logística.**

* Selección 1er Modelo: **8 variables seleccionadas por el método de MrMR**

In [19]:
# Asegúrate de que var_names está definido correctamente
selected_features = var_names[7]

# Añadiendo la constante al modelo
X_ent = sm.add_constant(x_ent[selected_features])
X_val = sm.add_constant(x_val[selected_features])

# Ajustando el modelo logístico
model = sm.Logit(y_ent, X_ent)
result = model.fit()

print("Variables seleccionadas:", selected_features)
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.476312
         Iterations 6
Variables seleccionadas: ['deu_ing', 'a_emp', 'n_educ', 'ing_h', 'a_dir', 'deu_o', 'edad', 'deu_tc']
                           Logit Regression Results                           
Dep. Variable:                default   No. Observations:                  350
Model:                          Logit   Df Residuals:                      341
Method:                           MLE   Df Model:                            8
Date:                Sun, 03 May 2026   Pseudo R-squ.:                  0.2003
Time:                        20:50:16   Log-Likelihood:                -166.71
converged:                       True   LL-Null:                       -208.47
Covariance Type:            nonrobust   LLR p-value:                 9.524e-15
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const  

* Selección 2do Modelo: **3 variables seleccionadas por poder estadístico**

In [20]:
selected_features = var_names[2]

# Añadiendo la constante al modelo
X_ent = sm.add_constant(x_ent[selected_features])
X_val = sm.add_constant(x_val[selected_features])

# Ajustando el modelo logístico
model = sm.Logit(y_ent, X_ent)
result = model.fit()

print("Variables seleccionadas:", selected_features)
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.486632
         Iterations 6
Variables seleccionadas: ['deu_ing', 'a_emp', 'n_educ']
                           Logit Regression Results                           
Dep. Variable:                default   No. Observations:                  350
Model:                          Logit   Df Residuals:                      346
Method:                           MLE   Df Model:                            3
Date:                Sun, 03 May 2026   Pseudo R-squ.:                  0.1830
Time:                        20:50:18   Log-Likelihood:                -170.32
converged:                       True   LL-Null:                       -208.47
Covariance Type:            nonrobust   LLR p-value:                 1.907e-16
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.0374      0.399     -5.105      0.0

**3. Evaluación Metodologica, parte 1**

* El **AUC** (Area Under the Curve) de la curva ROC (Receiver Operating Characteristic)  
* La **Matriz de Confusión**.  
* El **KS** (Kolmogorov-Smirnov)  
* La **Precisión**, **Recall** y **F1-Score**  

In [21]:
# Predicciones de probabilidad
y_pred_proba_ent = result.predict(X_ent)
y_pred_proba_val = result.predict(X_val)

# Umbral óptimo (por defecto 0.5, pero puedes optimizarlo)
threshold = 0.5
y_pred_ent = (y_pred_proba_ent >= threshold).astype(int)
y_pred_val = (y_pred_proba_val >= threshold).astype(int)

# F1-Score
f1_ent = f1_score(y_ent, y_pred_ent)
f1_val = f1_score(y_val, y_pred_val)

# KS
ks_ent = ks_statistic(y_ent, y_pred_proba_ent)
ks_val = ks_statistic(y_val, y_pred_proba_val)

# AUC
auc_ent = roc_auc_score(y_ent, y_pred_proba_ent)
auc_val = roc_auc_score(y_val, y_pred_proba_val)

print(f"F1-Score Entrenamiento: {f1_ent:.3f}")
print(f"F1-Score Validación: {f1_val:.3f}")
print(f"KS Entrenamiento: {ks_ent:.3f}")
print(f"KS Validación: {ks_val:.3f}")
print(f"AUC Entrenamiento: {auc_ent:.3f}")
print(f"AUC Validación: {auc_val:.3f}")

F1-Score Entrenamiento: 0.471
F1-Score Validación: 0.471
KS Entrenamiento: 0.461
KS Validación: 0.475
AUC Entrenamiento: 0.789
AUC Validación: 0.782


**4. Evaluación Metodologica, parte 2**: Diagnóstico de distribución de errores predictivos por variable 

Se realizó un análisis de densidad de variables por categoría de predicción para evaluar la capacidad de discriminación de las variables. El objetivo fue identificar en qué rangos de valor el modelo genera mayor certeza y en cuáles se producen los solapamientos que derivan en Falsos Positivos y Falsos Negativos.

In [22]:
df_ent['confusion_group'] = np.select(
    [
        (y_ent == 1) & (y_pred_ent == 1),  # TP
        (y_ent == 0) & (y_pred_ent == 0),  # TN
        (y_ent == 0) & (y_pred_ent == 1),  # FP
        (y_ent == 1) & (y_pred_ent == 0)   # FN
    ],
    ['TP', 'TN', 'FP', 'FN']
)

* Discriminación de `TP`, `TN`, `FP` y `FN` a través de gráficas de densidad por variables.

In [23]:
def plot_kde_by_confusion_group(
    df,
    features,
    group_col="confusion_group",
    order=("TN", "TP", "FP", "FN")
):
    var1, var2 = features[0], features[1]
    fig = make_subplots(rows=1, cols=2, subplot_titles=[var1, var2])
    colors = {'TN': 'blue', 'TP': 'green', 'FP': 'red', 'FN': 'orange'}
    for col_idx, var in enumerate([var1, var2], start=1):
        for g in order:
            data = df.loc[df[group_col] == g, var].dropna()
            kde_fig = ff.create_distplot(
                hist_data=[data],
                group_labels=[g],
                show_hist=False,
                show_rug=False,
                curve_type="kde"
            )
            fig.add_trace(
                go.Scatter(
                    x=kde_fig.data[0].x,
                    y=kde_fig.data[0].y,
                    name=g,
                    line=dict(color=colors[g]),
                    legendgroup=g,
                    showlegend=(col_idx == 1)  # leyenda solo en primer gráfico
                ),
                row=1, col=col_idx
            )
    fig.update_layout(
        title='Distribución de la variable por errores de clasificación',
        width=1200,
        height=600,
        paper_bgcolor='rgba(255,255,255,1)',
        plot_bgcolor='rgba(255,255,255,1)'
    )
    fig.update_yaxes(title_text='Densidad', col=1)
    fig.show()

In [24]:
plot_kde_by_confusion_group(df_ent, features=['deu_ing', 'edad'])

**Gráfico 1**

* Aquí se concentran los **TN** (azul) y los **FN** (rojo). Esto indica que cuando la relación deuda-ingreso es baja, el modelo tiende a predecir que el evento (default) no ocurrirá.

* Aquí se concentran los **TP** (naranja) y los **FP** (verde). Esto indica que cuando la deuda es alta respecto al ingreso, el modelo tiende a predecir que el evento (default) sí ocurrirá.

* **Nota.** Como hay un solapamiento significativo entre las 10 y 15 unidades de deu_ing, esa es la "zona de incertidumbre" donde el modelo comete la mayoría de sus errores de clasificación.

**5. Evaluación de Negocio.**

* **Paso 1.** De la muestra de Entrenamiento, se divide en 10 bucket la probabilidad estimada por el modelo.
* **Paso 2.** Se grafica cada uno de los 10 buckets vs la tasa de default observada en cada bucket promediando la variable `default`.
* **Paso 3.** Los mismo puntos de corte de cada bucket de la muestra de Entrenamiento, se aplican a la muestra de Validación y se grafica cada uno de los 10 buckets vs la tasa de default observada en cada bucket promediando la variable `default` pero en la muestra de Validación.

In [25]:
def distribucion_score(df_train, df_val, n_segments=10, score_col='SCORE', target_col='default', plot_graph=False):
    import plotly.graph_objects as go
    # Crear intervalos y etiquetas para los scores usando quantiles en el DataFrame de entrenamiento
    bins = pd.qcut(df_train[score_col], q=n_segments, retbins=True, duplicates='drop')[1]
    labels = [f"{round(bins[i], 2)} - {round(bins[i+1], 2)}" for i in range(len(bins)-1)]

    # Asignar los intervalos a los DataFrames
    df_train = df_train.copy()
    df_val = df_val.copy()
    df_train['Perfil'] = pd.cut(df_train[score_col], bins=bins, labels=labels, include_lowest=True)
    df_val['Perfil'] = pd.cut(df_val[score_col], bins=bins, labels=labels, include_lowest=True)

    # Resumen para entrenamiento
    resumen_train = df_train.groupby('Perfil').agg(
        Total=('Perfil', 'size'),
        Promedio_Default=(target_col, 'mean')
    ).reset_index()

    # Resumen para validación
    resumen_val = df_val.groupby('Perfil').agg(
        Total=('Perfil', 'size'),
        Promedio_Default=(target_col, 'mean')
    ).reset_index()

    # Generar gráficos si plot_graph es True
    if plot_graph:
        # Gráfico para el conjunto de entrenamiento
        fig_train = go.Figure()
        fig_train.add_trace(go.Bar(x=resumen_train['Perfil'], y=resumen_train['Total'], name='Total - Train', marker_color='indigo', yaxis='y1'))
        fig_train.add_trace(go.Scatter(x=resumen_train['Perfil'], y=resumen_train['Promedio_Default'], name='Probabilidad de Default - Train', marker=dict(color='orange', symbol='circle'), mode='lines+markers', yaxis='y2'))
        fig_train.update_layout(
            title='Entrenamiento',
            xaxis=dict(title='Perfil'),
            yaxis=dict(title='Total de Observaciones - Train', titlefont=dict(color='indigo')),
            yaxis2=dict(title='Probabilidad de Default - Train', titlefont=dict(color='orange'), overlaying='y', side='right'),
            legend=dict(orientation='h', yanchor='bottom', y=1.1, xanchor='center', x=0.5),
            width=1000, height=500,
            template='simple_white'
        )
        fig_train.show()

        # Gráfico para el conjunto de validación
        fig_val = go.Figure()
        fig_val.add_trace(go.Bar(x=resumen_val['Perfil'], y=resumen_val['Total'], name='Total - Validation', marker_color='teal', yaxis='y1'))
        fig_val.add_trace(go.Scatter(x=resumen_val['Perfil'], y=resumen_val['Promedio_Default'], name='Probabilidad de Default - Validation', marker=dict(color='red', symbol='circle'), mode='lines+markers', yaxis='y2'))
        fig_val.update_layout(
            title='Validación',
            xaxis=dict(title='Perfil'),
            yaxis=dict(title='Total de Observaciones - Validation', titlefont=dict(color='teal')),
            yaxis2=dict(title='Probabilidad de Default - Validation', titlefont=dict(color='red'), overlaying='y', side='right'),
            legend=dict(orientation='h', yanchor='bottom', y=1.1, xanchor='center', x=0.5),
            width=1000, height=500,
            template='simple_white'
        )
        fig_val.show()

    return resumen_train, resumen_val

* Lo que buscamos en estas graficias que el modelo logre dismcriminar los mejores perfiles (menores bucket vs menores tasas de default observadas) a mayor probabilidad de incumplimiento observada, es decir, que la probabilidad de incumplimiento sea monotonicamente creciente a mayores niveles de buckets.

In [26]:
df_ent['SCORE_logistica'] = y_pred_proba_ent
df_val['SCORE_logistica'] = y_pred_proba_val
dist_ent, dist_val = distribucion_score(df_ent, df_val, n_segments=10, score_col='SCORE_logistica', plot_graph=True)

Podemos agrupar algunos perfiles (siempre observando gráfico de entrenamiento):

1. **Nuevo Perfil 1**: Perfil 1 con Perfil 2 (igual probabilidad).
2. **Nuevo Perfil 2**: Perfil 3 con Perfil 4 (para atenuar PD del perfil 3).
3. **Nuevo Perfil 3**: Perfil 5 con Perfil 6 
4. **Nuevo Perfil 4**: Perfil 7 con perfil 8.
5. **Nuevo Perfil 5**: Perfil 9.
6. **Nuevo Perfil 6**: Perfil 10.

observar cómo queda, y si no hay monotonicidad se pueden agrupar de otra forma. 

## 5.2. Random Forest Classifier

Random Forest es un algoritmo de aprendizaje de conjunto (ensemble) basado en la combinación de múltiples árboles de decisión. Cada árbol se entrena de forma independiente y el resultado final se obtiene por votación mayoritaria (clasificación) o promedio (regresión).

Se apoya en dos principios fundamentales:

1. **Bagging**: cada árbol se entrena con una muestra aleatoria con reemplazo del dataset original

2. **Aleatoriedad de variables**: en cada nodo, solo se evalúa un subconjunto aleatorio de variables para encontrar el mejor corte

En esta sección revisaremos:

* 1. Ejecución del método de MrMR para selección de variables.
* 2. Entrenamiento del modelo de Random Forest Classifier.
* 3. Evaluación Metodologica, parte 1: métrica KS, F1, AUC, entre otras.
* 4. Evaluación de Negocio.


**1. Ejecución del método de MrMR para selección de variables.**

In [27]:
max_vars = len(x_ent.columns.tolist())
rf_model = RandomForestClassifier(n_estimators=10, criterion = 'entropy', max_depth = 5, min_samples_split = 10, random_state=42)
ks_ent_vars_rf, ks_val_vars_rf, var_names_rf = ks_vars(rf_model, x_ent, y_ent, x_val, y_val, max_vars=max_vars, graph=True, subtitulo='Random Forest')

100%|██████████| 10/10 [00:02<00:00,  4.72it/s]


* Los algoritmos de Random Forest utilizan la importancia para clasificiar las variables de mayor a menor importancia dentro de la descriminación de la variable default.

In [28]:
# Importancia de variables con Random Forest (barras verticales)
selected_features_rf = var_names_rf[7]  # Selecciona el mismo número de variables que en el ejemplo anterior

# Entrenamiento del modelo con las variables seleccionadas
rf_model_imp = RandomForestClassifier(n_estimators=10, criterion = 'entropy', max_depth = 5, min_samples_split = 10, random_state=42)
rf_model_imp.fit(x_ent[selected_features_rf], y_ent)

# Importancia de las variables
importancias = rf_model_imp.feature_importances_

df_importancias = pd.DataFrame({'Variable': selected_features_rf, 'Importancia': importancias})
df_importancias = df_importancias.sort_values(by='Importancia', ascending=False)
print(df_importancias)

# Gráfico de importancia (barras verticales)
fig = px.bar(df_importancias, x='Variable', y='Importancia', title='Importancia de Variables - Random Forest', color='Importancia', color_continuous_scale='Blues', orientation='v')
fig.update_layout(xaxis_title='Variable', yaxis_title='Importancia', bargap=0.2)
fig.show()

  Variable  Importancia
0  deu_ing     0.246221
5    deu_o     0.183109
1    a_emp     0.182547
4    a_dir     0.114003
7   deu_tc     0.089184
6     edad     0.070958
2   n_educ     0.058006
3    ing_h     0.055972


**3. Evaluación Metodologica, parte 1: métrica KS, F1, AUC, entre otras.**

In [29]:
# Predicciones y métricas con Random Forest
y_pred_proba_ent_rf = rf_model_imp.predict_proba(x_ent[selected_features_rf])[:, 1]
y_pred_proba_val_rf = rf_model_imp.predict_proba(x_val[selected_features_rf])[:, 1]

# Umbral óptimo (por defecto 0.5, pero puedes optimizarlo)
threshold = 0.5
y_pred_ent_rf = (y_pred_proba_ent_rf >= threshold).astype(int)
y_pred_val_rf = (y_pred_proba_val_rf >= threshold).astype(int)

# F1-Score
f1_ent_rf = f1_score(y_ent, y_pred_ent_rf)
f1_val_rf = f1_score(y_val, y_pred_val_rf)

# KS
ks_ent_rf = ks_statistic(y_ent, y_pred_proba_ent_rf)
ks_val_rf = ks_statistic(y_val, y_pred_proba_val_rf)

# AUC
auc_ent_rf = roc_auc_score(y_ent, y_pred_proba_ent_rf)
auc_val_rf = roc_auc_score(y_val, y_pred_proba_val_rf)

print(f"F1-Score Entrenamiento: {f1_ent_rf:.3f}")
print(f"F1-Score Validación: {f1_val_rf:.3f}")
print(f"KS Entrenamiento: {ks_ent_rf:.3f}")
print(f"KS Validación: {ks_val_rf:.3f}")
print(f"AUC Entrenamiento: {auc_ent_rf:.3f}")
print(f"AUC Validación: {auc_val_rf:.3f}")

F1-Score Entrenamiento: 0.690
F1-Score Validación: 0.340
KS Entrenamiento: 0.715
KS Validación: 0.525
AUC Entrenamiento: 0.918
AUC Validación: 0.783


* **Tarea**. Importante aplicar Grid Search para optimizar los hiperparámetros del modelo, como el número de árboles (`n_estimators`), la profundidad máxima de los árboles (`max_depth`), el número mínimo de muestras por hoja (`min_samples_leaf`), entre otros.

**4. Evaluación de Negocio.**

* Tiende a ordenar mejor los perfiles de riesgo vs probabilidad de incumplimiento observada, siendo que tienen las mismas variables que califican en el modelo final de regresión logística.

In [30]:
# Uso de la función para Random Forest
df_ent['SCORE_rf'] = y_pred_proba_ent_rf
df_val['SCORE_rf'] = y_pred_proba_val_rf
dist_ent_rf, dist_val_rf = distribucion_score(df_ent, df_val, n_segments=10, score_col='SCORE_rf', plot_graph=True)

In [31]:
df_ent

,default,edad,n_educ,a_emp,a_dir,ing_h,deu_ing,deu_tc,deu_o,flag_casado_missing,a_exp_missing,confusion_group,SCORE_logistica,SCORE_rf
190,1,38,3.0,12,14.0,63.0,16.0,1.138574,4.36,0,0,FN,0.360954,0.331296
172,0,35,1.0,9,1.0,34.0,5.0,0.400000,1.30,0,0,TN,0.102175,0.094822
251,0,23,2.0,0,4.0,21.0,8.7,0.450000,1.37,0,0,TN,0.399820,0.493694
77,0,52,1.0,16,25.0,64.0,8.6,1.870000,3.64,1,1,TN,0.076882,0.106853
134,1,29,3.0,3,3.0,50.0,2.4,0.260000,0.94,0,0,FN,0.196196,0.576022
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
171,1,26,2.0,2,6.0,24.0,13.6,1.580000,1.69,0,0,TP,0.509964,0.812804
56,0,44,1.0,12,3.0,31.0,9.7,0.360000,2.65,0,0,TN,0.133475,0.147296
49,0,26,1.0,4,3.0,20.0,5.4,0.320000,0.76,0,0,TN,0.176615,0.169978
234,0,31,2.0,9,1.0,30.0,5.8,0.970000,1.14,0,0,TN,0.136045,0.144211


---
# 6. Tarea

* 1. Use otros algortimos: XGBoost, LightGBM, CatBoost, entre otros y compare los resultados con el modelo de regresión logística (**modelo benchmark**).

* 2. Seleccione un modelo, calcule: **(1).** Calibración y **(2).** Estabilidad de la probabilidad de incumplimiento estimada por el modelo usando PSI (Population Stability Index) entre la muestra de entrenamiento y validación.

* 3. Luego aplique caso de negocio en excel con las probabilidades entregadas por el modelo y los datos entregados del excel.

**Recuerde** que si selecciono un modelo tendrán que ahora ejecutar el mismo modelo con los mismo parametros pero con toda la muestra (sin separar entre entrenamiento y validación) para que el modelo aprenda de los datos que dejo fuera en la muestra de validación.

---